cmd/libsnap-confine-private/infofile.c

```diff
@@ -0,0 +1,215 @@
+ * key=value pairs and passing them along to the scanner function.
+ *
+ * The stream is scanned exactly once, using internally managed buffer. The
+ * buffer is reused as key/value storage for the purpose of the scanner
+ * function. The values provided to the scanner function must be copied if they
+ * need to outlive the lifetime of the call into the scanner.
+ *
+ * Each line must be of the format key=value where key and value are arbitrary
+ * strings, excluding the NUL byte which would be confusing in traditional C
+ * strings.
+ *
+ * On success the return value is zero and err_out, if not NULL, is deferences
+ * and set to NULL.  On failure the return value is -1 is and detailed error
+ * information is stored by dereferencing err_out.  If an error occurs and
+ * err_out is NULL then the program dies, printing the error message.
+ **/
+static int sc_infofile_scan(FILE *stream, sc_infofile_scanner_fn scanner_fn, void *caller_state, sc_error **err_out);
+
+/**
+ * sc_infofile_get_key_state represents caller state for sc_infofile_get_key.
+ *
+ * The state gets passed to sc_infofile_scan and is used to convey the key
+ * that is being looked for as well as the value that was found.
+ **/
+typedef struct sc_infofile_get_key_state {
+    const char *wanted_key;
+    char *stored_value;
+} sc_infofile_get_key_state;
+
+/**
+ * sc_infofile_get_key_scanner is the scanner callback for sc_infofile_get_key.
+ *
+ * The callback unpacks the scanner state and if a key is found, stores the value
+ * into the caller state and stops the scanning process.
+ **/
+static int sc_infofile_get_key_scanner(sc_infofile_scanner_state *scanner_state, sc_error **err_out) {
+    sc_error *err = NULL;
+    if (scanner_state == NULL) {
+        err = sc_error_init(SC_LIBSNAP_ERROR, SC_BUG, "scanner_state cannot be NULL");
+        goto out;
+    }
+    sc_infofile_get_key_state *caller_state = scanner_state->caller_state;
+    if (caller_state == NULL) {
+        err = sc_error_init(SC_LIBSNAP_ERROR, SC_BUG, "caller_state cannot be NULL");
+        goto out;
+    }
+
+    if (sc_streq(caller_state->wanted_key, scanner_state->key)) {
+        caller_state->stored_value = sc_strdup(scanner_state->value);
+        scanner_state->stop = true;
+    }
+
+out:
+    return sc_error_forward(err_out, err);
+}
+
+int sc_infofile_get_key(FILE *stream, const char *key, char **value, sc_error **err_out) {
+    sc_error *err = NULL;
+
+    /* NOTE: stream is checked by sc_infofile_scan */
+    if (key == NULL) {
+        err = sc_error_init(SC_LIBSNAP_ERROR, SC_API_MISUSE, "key cannot be NULL");
+        goto out;
+    }
+    if (value == NULL) {
+        err = sc_error_init(SC_LIBSNAP_ERROR, SC_API_MISUSE, "value cannot be NULL");
+        goto out;
+    }
+
+    sc_infofile_get_key_state scanner_state = {.wanted_key = key};
+    sc_infofile_scanner_fn scanner_fn = sc_infofile_get_key_scanner;
+    /* Call the scanner and ignore the error code. We will forward any
```

Comment:
This could probably do without the comment:
```
*value = NULL // initialize with meaningful value
if (sc_infofile_scan(stream, scanner_fn, &scanner_state, &err) == 0) {
    *value = scanner_state.store_value;
}
..
```

Reply:
Yeah, that would work too.

Reply:
Did something similar now.